pkg/publisher/fallback/publisher.go

```diff
@@ -0,0 +1,78 @@
+// fallback implements a publisher.Publisher that will try multiple other
+// Publishers in order until one suceeeds.
+//
+// The Publishers are tried in the order specified in the call to
+// NewFallbackPublisher. If and only if all of the Publishers return an error
+// result, the fallback publisher will also return an error result. Otherwise,
+// it will return the result of the first Publisher to succeed and will swallow
+// any errors. Subsequent Publishers will not be attempted after one succeeds.
+
+package fallback
+
+import (
+	"context"
+
+	"github.com/filecoin-project/bacalhau/pkg/model"
+	"github.com/filecoin-project/bacalhau/pkg/publisher"
+	"github.com/rs/zerolog/log"
+	"go.uber.org/multierr"
+)
+
+type fallbackPublisher struct {
+	publishers []publisher.Publisher
+}
+
+func NewFallbackPublisher(publishers ...publisher.Publisher) *fallbackPublisher {
+	return &fallbackPublisher{
+		publishers: publishers,
+	}
+}
+
+// fallback accepts a slice of publisher objects and passes them to the supplied
+// function in order, until one does not return an error value and then returns
+// that result. If all publishers return an error value, a composite error is
+// returned.
+func fallback[T any, P any](ctx context.Context, publishers []P, method func(P) (T, error)) (T, error) {
+	var anyErr error
+	for _, publisher := range publishers {
+		result, err := method(publisher)
+		if err == nil {
+			return result, nil
+		} else {
+			log.Ctx(ctx).Warn().Msgf("publisher %v returned an error: %s", publisher, err.Error())
+			anyErr = multierr.Append(anyErr, err)
+		}
+	}
+
+	var zeroResult T
+	return zeroResult, anyErr
+}
+
+// ComposeResultReferences implements publisher.Publisher
+func (f *fallbackPublisher) ComposeResultReferences(ctx context.Context, jobID string) ([]model.StorageSpec, error) {
+	return fallback(ctx, f.publishers, func(p publisher.Publisher) ([]model.StorageSpec, error) {
```

Comment:
what if we succeed with `PublishShardResult` using publisher A and then when we try to `ComposeResultReferences` and publisher A fails - are we now trying to `ComposeResultReferences` using publisher B but for results that were generated by publisher A?

Reply:
> This perhaps means ComposeResultReferences shouldn't be a part of this interface? And is worth a refactor?

I was just thinking about this and I think actually it's OK because the publish result is a `StorageSpec` - and those should be capable of downloading results from whereever they live (it's the point of a storage spec after all)

So - I'm not sure we actually do need to know "which pubisher published" - as long as the act of publishing returns a storage spec that can be resolved then it's OK

For example - in the Estuary fallback case - it's either on IPFS or on IPFS :-)